blueman/gui/applet/PluginDialog.py

```diff
@@ -159,45 +169,96 @@ def __init__(self, applet: "BluemanApplet") -> None:
         self.connect("delete-event", self._on_close)
 
-        self.list.set_cursor(Gtk.TreePath.new_first())
-
         close_action = Gio.SimpleAction.new("close", None)
         close_action.connect("activate", lambda x, y: self.close())
+
         self.add_action(close_action)
 
-    def list_compare_func(self, _treemodel: Gtk.TreeModel, iter1: Gtk.TreeIter, iter2: Gtk.TreeIter, _user_data: object
-                          ) -> int:
-        a = self.list.get(iter1, "activatable", "name")
-        b = self.list.get(iter2, "activatable", "name")
+    def _add_plugin_action(self, name: str, state: bool, activatable: bool) -> None:
+        logging.debug(f"adding action: {name}")
+        action = Gio.SimpleAction.new_stateful(
+            name, None, GLib.Variant.new_boolean(state)
+        )
+        action.set_property("enabled", activatable)
+        self.add_action(action)
+        action.connect("change-state", self._on_plugin_toggle)
+
+    def _widget_factory(self, item: GObject.Object, _data: Optional[object] = None) -> Gtk.Widget:
+        assert isinstance(item, PluginItem)
+        box = Gtk.Box(orientation=Gtk.Orientation.HORIZONTAL, spacing=5, visible=True)
+
+        checkbutton = Gtk.CheckButton(visible=True, sensitive=item.props.activatable)
+        box.add(checkbutton)
+        checkbutton.set_action_name(f"win.{item.props.plugin_name}")
+
+        self._add_plugin_action(item.props.plugin_name, item.props.enabled, item.props.activatable)
+        # Set active after adding action
+        checkbutton.set_active(item.props.enabled)
 
-        if (a["activatable"] and b["activatable"]) or (not a["activatable"] and not b["activatable"]):
-            if a["name"] == b["name"]:
+        plugin_im = Gtk.Image(icon_name=item.props.icon_name, visible=True)
+        box.add(plugin_im)
+
+        label = Gtk.Label(label=item.props.description, use_markup=True, visible=True)
+        box.add(label)
+        return box
+
+    def _model_sort_func(self, item1: Optional[object], item2: Optional[object], _data: Optional[object] = None) -> int:
+        assert isinstance(item1, PluginItem)
+        assert isinstance(item2, PluginItem)
+
+        if (item1.props.activatable and item2.props.activatable) or (not item1.props.activatable and
+                                                                     not item2.props.activatable):
+            if item1.props.plugin_name == item2.props.plugin_name:
                 return 0
-            if a["name"] < b["name"]:
+            if item1.props.plugin_name < item2.props.plugin_name:
                 return -1
-            else:
-                return 1
-
+            return 1
         else:
-            if a["activatable"] and not b["activatable"]:
+            if item1.props.activatable and not item2.props.activatable:
                 return -1
-            elif not a["activatable"] and b["activatable"]:
+            if not item1.props.activatable and item2.props.activatable:
```

Comment:
Sonarcube complains that `not item1.props.activatable` is always `True` here. Makes some sense, but fine for me to leave the structure. :shrug:

Reply:
I didn't change (or think about) the logic but I'll have a look to see if this can be handled better.

Reply:
I think it makes sense to put these two checks for different `activatable` values first, then the two for different `plugin_name`s and finally a single `return 1` and save what's currently in lines 209 and 210. There should not be any redundancy as caught by Sonarcube then.

Reply:
Pushed update that works for me and it's happy now.